### ISSUE TITLE
pgxpool queries hanging when using gin / concurrency

### ISSUE THREAD

**Issue by matthew-graves** [jackc/pgx#1126]
Hello,

I wasn't sure where else to put this information up for it to be shared for other people to find so I'm dropping it in this issue for now. I wanted to add a note about this to documentation about concurrency, and I don't mind submitting a PR if there's an appropriate place to put this info. 

I had an issue with queries hanging when using Gin + pgxpool but this applies to all concurrent functions using pgxpool. I couldn't figure out why my containers that use gin + pgxpool would inevitably hang after a certain period of usage. I'd put in error handling for every scenario I could think of, but any endpoint that would require a query would end up hanging indefinitely with no errors.

The problem stemmed from me only using `defer rows.Close()` on my functions that executed multiple queries. 

Because the connection does not get closed when using defer until the function ends, if another function uses up all other available connections from pgxpool before the first function can finish, both functions (and all future ones) end up deadlocked. This is because they cannot obtain a new connection to complete the remaining queries in the function, and release the existing connections back to the pool. 

The easy fix for this, is to add in a rows.Close() after the data is finished being scanned and read. Per the documentation, if it's already closed, it's a no-op function so theoretically no problems by implementing this at the appropriate place in the code, and leaving the defer rows.close() will still release any connections that are caught and returned by errors. 

tl;dr

Concurrent functions that use multiple queries should explicitly close connections after reading data, in addition to a deferred closing of the connection for safety. 

If I'm completely off base with this, please feel free to correct me. 

Hope this helps someone in the future as I'd lost a couple nights sleep over tracking this down.

**Comment by ysomad** [comment]
hello! I'm using gin + pgxpool too, directly calling `pool.QueryRow(...).Scan(...)` and my sql request hangs on `maxPoolSize + 1` attempt. Maybe you know why? 

In this case, `Scan()` closes rows.

**Comment by thinkingatoms** [comment]
> hello! I'm using gin + pgxpool too, directly calling `pool.QueryRow(...).Scan(...)` and my sql request hangs on `maxPoolSize + 1` attempt. Maybe you know why?
> 
> In this case, `Scan()` closes rows.

hi, I'm using gochi + pgxpool and after some simple queries against stored procedures, pgxpool just hangs.  Did you ever solve your issue?  Thanks in advance!

**Comment by ysomad** [comment]
> > hello! I'm using gin + pgxpool too, directly calling `pool.QueryRow(...).Scan(...)` and my sql request hangs on `maxPoolSize + 1` attempt. Maybe you know why?
> > In this case, `Scan()` closes rows.
> 
> hi, I'm using gochi + pgxpool and after some simple queries against stored procedures, pgxpool just hangs. Did you ever solve your issue? Thanks in advance!

greetings, yes I did. I had a problem when I used QueryRow/Query and I didn't read from it (when they didn't return anything but used scan). Changed to Exec where needed and that how I fixed the hang.

**Comment by thinkingatoms** [comment]
> > > hello! I'm using gin + pgxpool too, directly calling `pool.QueryRow(...).Scan(...)` and my sql request hangs on `maxPoolSize + 1` attempt. Maybe you know why?
> > > In this case, `Scan()` closes rows.
> > 
> > 
> > hi, I'm using gochi + pgxpool and after some simple queries against stored procedures, pgxpool just hangs. Did you ever solve your issue? Thanks in advance!
> 
> greetings, yes I did. I had a problem when I used QueryRow/Query and I didn't read from it (when they didn't return anything but used scan). Changed to Exec where needed and that how I fixed the hang.

thanks so much for getting back to me!  yes after hours last night i realized my Query calls that returned Rows object needed a "defer rows.Close()".  (I was reading/scanning with rows but was still losing the connection per call).  very much appreciate your pointer!
